fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/properties/FDBPropertyKey.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * FDBPropertyKey.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb.properties;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.function.Supplier;
+
+@SuppressWarnings("unused")
+/**
+ * A key for a property that adopter configures for Record Layer.
+ * The value is lazily loaded from the supplier when the property is read.
+ * If no supplier is provided, a default value is used.
+ */
+public final class FDBPropertyKey<T> {
+    @Nonnull
+    private final String name;
+    @Nonnull
+    private final T defaultValue;
+    @Nonnull
+    private final Class<T> type;
+    @Nullable
+    private final Supplier<T> valueSupplier;
+    @Nullable
+    private T cachedValueFromSupplier;
+
+    private FDBPropertyKey(@Nonnull final String name, @Nonnull T defaultValue,
+                           @Nonnull Class<T> type, @Nullable final Supplier<T> valueSupplier) {
+        this.name = name;
+        this.defaultValue = defaultValue;
+        this.type = type;
+        this.valueSupplier = valueSupplier;
+    }
+
+    /**
+     * The name for each property needs to be unique and in a reverse FQDN format according to the package it is located.
+     *
+     * @return the name of this property
+     */
+    @Nonnull
+    public String getName() {
+        return name;
+    }
+
+    @Nonnull
+    public Class<T> getType() {
+        return type;
+    }
+
+    @Nonnull
+    public T getValue() {
+        if (valueSupplier == null) {
+            return defaultValue;
+        } else if (cachedValueFromSupplier != null) {
+            return cachedValueFromSupplier;
```

Comment:
After an offline discussion with @alecgrieser , the cached value is removed. A cached value may be helpful if one prop's value is loaded multiple times by RL, but on the other hand it could be a problem if a prop a dynamic one and has a longer lifecycle. This could be a topic for further discussion in future, like how we want to support dynamic prop object with long lifecycle and also avoid loading the same value multiple times. But none of them is related to the current use case. So I just removed it for now.